Update AI test plans for possible full support

fedinskiy opened the pull request:
### Links

JIRA: https://issues.redhat.com/browse/QUARKUS-6258

Quarkus documentation: https://quarkus.io/blog/quarkus-meets-langchain4j/

Extension/feature community status: preview

### Reminder for considerable topics

 - [x] Make sure you have considered the following areas when preparing the test plan:

   - Logging
   - Tracing
   - Metrics
   - Security
   - OpenAPI
   - Data sources
   - Frontend
   - Qute
   
   
Follow-up to https://github.com/quarkus-qe/quarkus-test-plans/pull/241

jcarranzan commented on QUARKUS-6258.md:
```diff
@@ -15,9 +15,16 @@ The list of prioritized features (taken from the issue):
+We also need to verify correct usage of OpenAI and Watson.x (optional for the first release) APIs as well as observability (metrics, logging, and telemetry), the latter being a requirement for full support.
+
+### Optional test cases
+These are required only for test-preview status in case Quarkus will go the way of maximum support in te first release:
```
```suggestion
These are required only for test-preview status in case Quarkus will go the way of maximum support in the first release:
```

jcarranzan reviewed (commented):
I will need to catch up more on this AI coverage topic, but I've left a minor suggestion.
On the other hand, maybe we should just focus on OpenAI in this TP and leave Watson.X for a future next test plan?

michalvavrik commented on QUARKUS-6258.md:
```diff
@@ -47,6 +54,10 @@ Primary focus will be on RHEL-based testing to ensure functionality of proposed
 As we aim for tech preview level for 3.27, we won't be focusing on full platform matrix coverage. We will be gaining experience to expand the coverage and platforms in the upcoming releases.d testing.
 
+#### Impact on test suites for Full Support
+Since Quarkus AI may be moved to full support in its first release (Quarkus 3.27), we need to have a full support on all platforms. Tha means, that all tests should be in Quarkus Test suite, to guarantee running an all supported platforms.
```
What is it "full support on all platforms"? RHEL, OpenShift, Windows, IBM Z/PE..? I thought that this is not intended for be RHBQ, hence the platforms may differ to RHBQ. Why are you not more specific, maybe you write it somewhere else? Because in the `Impact on test suite` you mention RHEL and OCP.

jedla97 replied:
So just to clarify, if the full support is decided, the tests won't be added to quarkus-lang4j repository (upstream tests from the Quarkiverse extensions) but all the test will be Quarkus TS (I assume you mean Quarkus QE TS).

Or there will be some tests in quarkus-lang4j repository and some more complex/with extensions compatibility will be in QE TS

jedla97 replied:
Probably out of scope of this TP/strategy but support on OpenShift is from our side defined as deploying the Quarkus AI app and that app is able to communicating with OpenAi/Watson/Geminy etc.?

Is there plan to try the `Red Hat OpenShift AI` compatibility with Quarkus? I assume it's the similar as communicating with external API but additional step would be deploying own model. This is probably not needed for full support but I wouldn't be surprised if there would need for next LTS 3.X.

mocenas replied:
```suggestion
Since Quarkus AI may be moved to full support in its first release (Quarkus 3.27), we need to have a full support on all platforms. That means, that all tests should be in Quarkus Test suite, to guarantee running on all supported platforms.
```

michalvavrik commented on QUARKUS-6258.md:
```diff
@@ -47,6 +54,10 @@ Primary focus will be on RHEL-based testing to ensure functionality of proposed
 
 As we aim for tech preview level for 3.27, we won't be focusing on full platform matrix coverage. We will be gaining experience to expand the coverage and platforms in the upcoming releases.d testing.
 
+#### Impact on test suites for Full Support
```
On line 53 you write `Primary focus will be on RHEL-based testing to ensure functionality of proposed features against productized bits. For the OpenShift side, we will explore the possibility of using the @GitRepositoryQuarkusApplication approach.`, what does it mean "primary focus", does it mean you don't test anything else or how much? Such a soft metrics are acceptable, but it would be nice to mention how non-primary focus looks like?

michalvavrik replied:
why is "Impact on test suite" subsection of the "Existing test coverage"?

michalvavrik replied:
Okay, that is not how I understood "non-primary", but unless someone else have same issue, I am fine with keeping it as is. Thanks

michalvavrik commented on QUARKUS-6258.md:
```diff
@@ -15,9 +15,16 @@ The list of prioritized features (taken from the issue):
 - Advanced RAG
 - Guardrails 	
 - WebSocket chat integration 	
```
"WebSocket chat integration", this refers to the WS Next; what is actual scope? you will have one test? Or test for every relevant WS Next scenario (WS Client, WS Server, broadcasting, pings, uni/multi/non-reactive formats), use it to test every tool/service/model. The reason I am asking, how do you know what is relevant if you don't plan it, state it in this plan and give someone change to challenge that.

michalvavrik commented on QUARKUS-6258.md:
```diff
@@ -15,9 +15,16 @@ The list of prioritized features (taken from the issue):
+These are required only for test-preview status in case Quarkus will go the way of maximum support in te first release:
+- Scoring Model
+- Parsing of documents (PDF, Word, etc)
+- MCP client and server (STDIO and Streamable HTTP )
```
wow, can you elaborate what you mean by `MCP client and server (STDIO and Streamable HTTP )`? Because on testing Quarkus MCP server, you would have great deal of work, does it mean that you just want to test some scenarios that use MCP client/server?

michalvavrik commented on QUARKUS-6258.md:
```diff
@@ -15,9 +15,16 @@ The list of prioritized features (taken from the issue):
+- Scoring Model
+- Parsing of documents (PDF, Word, etc)
+- MCP client and server (STDIO and Streamable HTTP )
+- Testing AI infused application (required for dev preview)
```
what is the dev preview? I know tech-preview, I know dev support https://access.redhat.com/support/offerings/developer/. I think I have heard it, but I have problem to find it, is this the right term?

jedla97 replied:
Not 100% sure but it should be this https://access.redhat.com/support/offerings/devpreview as one of the support level defined in https://access.redhat.com/articles/7040491

michalvavrik replied:
I got answer in the meeting, thanks

michalvavrik reviewed (commented):
I have left couple of questions/comments, I am uneducated on this topic so take it easy. In general, I'd not know how to start review this TP  because it is so vague, I can't argue on anything. like "verify correct usage of OpenAI and Watson.x" what the hell does it mean? Where do I find what "correct usage" is, and so on, it is ever item of test cases.

We agreed that I am not going to review, so there is just one thing I'd like you to answer - what about security? it is not mentioned once in this TP, does it mean that you don't need to test it?

mjurc commented on QUARKUS-6258.md:
```diff
@@ -15,9 +15,16 @@ The list of prioritized features (taken from the issue):
+- CDI Chat Model
 
-We also need to verify correct usage of OpenAI and Watson.x (optional for the first release) APIs as well as observability (metrics, logging, and telemetry).
+We also need to verify correct usage of OpenAI and Watson.x (optional for the first release) APIs as well as observability (metrics, logging, and telemetry), the latter being a requirement for full support.
```
This is really vaguely worded. Let's be precise in technical writing, please.

What are the scenarios and what is the correct usage in those scenarios? What are the integration points?

fedinskiy replied:
> correct usage

The app can send queries to these APIs, receive answers and not fail or loose data in the process.

mjurc commented on QUARKUS-6258.md:
```diff
@@ -40,12 +47,16 @@ According to Mario, this is intentional, since they consider upstream lang4chain
+
+## Impact on test suites for Full Support
+Since Quarkus AI may be moved to full support in its first release (Quarkus 3.27), we need to have a full support on all platforms. This means, that all tests should be in Quarkus Test suite, to guarantee running an all supported platforms.
+Additionally, we should add verification for this extension to Marete and startstop tests.
```
Marete needs to be updated even for tech preview extensions. We test support levels of extensions.

As mentioned before, please clarify support level for extensions, and explicitly mention what needs to be done for acceptance testing.

What start-stop scenarios do you anticipate?

Let's be a bit more precise.

fedinskiy replied:
Support level is described in the upstream issue, wouldn't that be enough?

> What start-stop scenarios do you anticipate?
No idea. We run start-stop tests for other supported extensions, maybe we should do it for these ones as well.

mjurc commented on QUARKUS-6258.md:
```diff
@@ -40,12 +47,16 @@ According to Mario, this is intentional, since they consider upstream lang4chain
 
-### Impact on test suite
+## Impact on test suite
 We will focus on updating and running upstream tests from the Quarkiverse extensions against productised binaries.
```
Again, this is really vague.

The test plan misses
* what the scenarios do
* what are the pass criteria
* how will the upstream test suite be ran with RHBQ platform
* how long the jobs will run and what infra they require

For full support, this test plan also needs to account for the fact that LTS lives for a long time and the upstream needs some sort of way to check that the backports work properly without introducing regressions, and that the development branch of the extensions is still usable with Quarkus platform in development

mjurc commented on QUARKUS-6258.md:
```diff
@@ -40,12 +47,16 @@ According to Mario, this is intentional, since they consider upstream lang4chain
+Additionally, we should add verification for this extension to Marete and startstop tests.
 
 ## Impact on resources
 The tests themselves do not consume much resources (single ci.m1.medium/ci.m1.large should be enough).
```
We usually also have time estimation of test runtime.

This will also increase for full support, so let's have the accounting updated, and broken down per test suite.

mjurc replied:
Let's start with an accounting of what jobs are supposed to be changed/added, time impact can be added once you have an idea

mjurc commented on QUARKUS-6258.md:
```diff
@@ -40,12 +47,16 @@ According to Mario, this is intentional, since they consider upstream lang4chain
-As we aim for tech preview level for 3.27, we won't be focusing on full platform matrix coverage. We will be gaining experience to expand the coverage and platforms in the upcoming releases.d testing.
+As we aim for tech preview level for 3.27, we won't be focusing on full platform matrix coverage. We will be gaining experience to expand the coverage and platforms in the upcoming releases.
+
+## Impact on test suites for Full Support
```
For full support, this test plan also needs to account for the fact that LTS lives for a long time and the upstream needs some sort of way to check that the backports work properly without introducing regressions, and that the development branch of the extensions is still usable with Quarkus platform in development

fedinskiy replied:
Development branch runs CI with mocks, which should be enough to ensure compatibility with Quarkus.

Regarding backports and lack of regressions — will this CI be enough, or should we run these tests ourselves using real LLMs?

mjurc replied:
Well, I'd expect the tester to commit an analysis and to answer this question :)

This is a new area, so actually having an idea of what the challenges will be and some sort of decision matrix would be super useful to explain what drove the decisions for the testing rather than sticking to whatever comes and the first idea

fedinskiy replied:
Added regular running on upstream

mjurc commented on QUARKUS-6258.md:
```diff
@@ -64,10 +75,14 @@ Important notes:
 
+## Non-automated testing
+
+See https://github.com/quarkus-qe/quarkus-test-plans/blob/main/QUARKUS-6257.md#additional-requirements-for-full-support for details.
```
Why is this in non-automated testing? We need to test all those requirements automatically (except #7, which is an attestation) if this is to be fully supported (and we do for all the supported extensions already).

fedinskiy replied:
I doubt we can check upstream guides, existence of CVE detection by tools and confidence of dev team automatically.

mjurc replied:
I suppose that's right, but it's still really unclear considering most of those points are not automated

fedinskiy replied:
That is why I put these in non-automated testing section. Since we agree on that, am I correct to assume this matter to be resolved?

fedinskiy reviewed (commented):
@michalvavrik 

> We agreed that I am not going to review, so there is just one thing I'd like you to answer - what about security? it is not mentioned once in this TP, does it mean that you don't need to test it?
I am planning to check that the API keys are nto leaked, but that it. I do not see any other security implications of this feature, do you?

@mocenas 

> I would like to have more about how will we test on openshift. E.g. will only the app be running on OCP and connecting to LLM outside? Are there plans/possibilities (does it make sense?) to run some part of AI ecosystem in the OCP?

The answer is in this thread: https://github.com/quarkus-qe/quarkus-test-plans/pull/253#discussion_r2292166911

fedinskiy commented on QUARKUS-6258.md:
```diff
@@ -15,9 +15,16 @@ The list of prioritized features (taken from the issue):
 - Guardrails 	
 - WebSocket chat integration 	
-- CDI Chat Model.
+- CDI Chat Model
```
Added pass criteria and guides

fedinskiy commented:
@rsvoboda regarding your comment:

> Testing infrastructure
    
I expect the same as usual, OpenStack+AWS. Or do you suggest to specifically mention usage of API of OpenAI?

> Testing on all platforms

Added it to the section "Impact on test suites for Full Support"

>CLI updates reflect Quarkus LangChain4j

asked @geoand about it, he said, that this is not currently a priority

> Quarkus LangChain4j Extensions out of Preview level

please, elaborate

>Quarkus LangChain4j integration TS runnable against platform BOM

added this to " Impact on test suite"

>Breaking changes tracked in quarkus-langchain4j project

https://github.com/quarkiverse/quarkus-langchain4j/issues?q=is%3Aissue%20state%3Aopen%20label%3Arelease%2Fbreaking-change (@geoand said, they started to use it recently, so no issues so far)

>Release tracking in quarkus-langchain4j project similar to Quarkus
        https://quarkus.io/blog/tag/release/

Please elaborate. There is this page https://github.com/quarkiverse/quarkus-langchain4j/releases, is it not good enough?

Also, GitHub PRs are not really convenient place to have long discussions, so let's have them in slack/jira next time.

mjurc commented on QUARKUS-6258.md:
```diff
@@ -38,14 +49,19 @@ Given the above I do not recommend to (re)use this repository for product valida
+## Impact on test suites for Full Support
+Since Quarkus AI may be moved to full support in its first release (Quarkus 3.27), we need to have a full support on all (see below) platforms. This means, that all tests should be in Quarkus Test suite, to guarantee running an all supported platforms.
+Additionally, we should add verification for this extension to Marete and startstop tests.
+Note: Requirement about "all platforms" (without further details) came from the PM. Quarkus QE runs tests for RHEL, Openshift and Windows, so this is the platforms we will cover Quarkus AI on. While Quarkus is also supported on IBM Z& IBM P, it is out of scope for Quarkus QE team (there is a separate IBM team, which handles it). 
```
I'm not really sure what this sentence means, or why is it important for test plan.

But if an extension or feature is declared to be fully supported, it must work on all the certified platforms. That's the definition. That's not what the reviewed sentence says, though.

michalvavrik replied:
Yes, I requested to explain what "all supported platforms" mean. Is it just me who see a difference between "all supported platforms" meaning "platforms that we will support for AI" and all supported platforms" meaning "all platforms we support for RHBQ"? You are going to support Windows and OpenShift and Aarch64, cool. You can just tweak that text to provide clear meaning so that it can be verified. e.g.

```
[all supported platforms](https://access.redhat.com/articles/4966181)
```

is clear to me

mjurc commented on QUARKUS-6258.md:
```diff
@@ -6,18 +6,29 @@ Verify features of Quarkus AI, which are going to be added in phase 1
+
+We also need to verify that Quarkus AI apps can work properly with both OpenAI and Watson.x APIs (the latter is optional for the first release) and that it is integrated with observability (generate metrics, logs, and telemetry).
+
+Pass criteria: for each of these features there is at least one test, which uses it to connect to LLM and get an answer. The answer looks like the query (and its context) was passed to LLM without losing data and the answer was retrieved without losing data either.
```
Do we maybe have any security concerns here?

michalvavrik replied:
> I do not think so. Quarkus AI is a convenience wrapper around rest client. The worst thing it can do, is to connect to the wrong URI and leak the credentials. No sure if we need to check such specific cases.

What about prompt injection, are there tests for it? Is handling of generative AI model output secure? Do you test that Quarkus properly sanitize AI model output? If using some tools require authorization (like cancelling booking will require me to actually have admin account or be that user who created it), do standard security annotations work in these contexts? What about communication with AI models, I assume these data can be highly sensitive, are you testing TLS, what authentication protocols are you going to test?

For example here are top 10 risks for LLMs https://genai.owasp.org/llm-top-10/, aren't some of them relevant for your testing?

michalvavrik replied:
>  Please, elaborate. Usually, it is input, that is being sanitized.
> Sanitisation of output

https://genai.owasp.org/llmrisk/llm052025-improper-output-handling/ I was thinking such a scenarios where Quarkus deals with the output, that is more than "pass the string down to the user code". For example sentence `LLM output is used to construct file paths without proper sanitization, potentially resulting in path traversal vulnerabilities.` can user expect `Path` class from LLM or do they have to get `String` and create the `Path` object.

> The looks through the list shows, that these are mostly vulnerabilities of LLMs or theirs usage, not really applicable to our case of middleware.

I thought LLMs are allowed to call tools, like here https://docs.quarkiverse.io/quarkus-langchain4j/dev/quickstart-function-calling.html#_step_1_create_the_toolbox, which means that not every LLMs output is user responsibility, also Quarkus does some actions with it? If so:

1. If the LLMs output is processed (in the example is it `location`? that's a question, I don't use it unlike you). AFAICT tool arg can be String, long https://github.com/quarkiverse/quarkus-langchain4j/blob/a917562fb00ce64ac52d58c2cef0631f7e95fba6/core/deployment/src/test/java/io/quarkiverse/langchain4j/test/ToolSpecificationTest.java#L47 or even some DTO https://github.com/quarkiverse/quarkus-langchain4j/blob/a917562fb00ce64ac52d58c2cef0631f7e95fba6/core/deployment/src/test/java/io/quarkiverse/langchain4j/test/ToolSpecificationTest.java#L47 . Should you check what happens when tool args format is incorrect, is it an issue?
2. https://docs.quarkiverse.io/quarkus-langchain4j/dev/quickstart-function-calling.html#_step_2_register_the_ai_service says that `@ToolBox` specifies which tools is the service allowed to call. Don't you need to check that the service cannot call a tool outside of the toolbox?

mjurc reviewed (changes requested):
There's some improvements, but I still think it's really vague with a lot of room for interpretation. That's really not good for any plan :(

I also think it would be much clearer for readers to understand if there was actual plan on how we achieve tech preview support, and after that a section on what needs to be done to achieve full support

mjurc reviewed (changes requested):
The discussions have been resolved but the document is still very vague and has grown out of date. It needs to be updated with current status, and the actual plan needs to be written.

fedinskiy commented on QUARKUS-6258.md:
```diff
@@ -38,19 +52,25 @@ Given the above I do not recommend to (re)use this repository for product valida
-### Impact on test suite
-We will focus on updating and running upstream tests from the Quarkiverse extensions against productised binaries.
+## Impact on test suite
+Given requirement to run tests with platform BOMs, we should focus on adding tests to Quarkus QE TS. Modules in Quarkiverse `intergration-tests` and `samples` folder can be used a source of inspiration, or, in the latter case, used directly with a help of `@GitRepositoryQuarkusApplication`.
```
> Is this still the plan for 3.27
It is.
>It seems a bit out of date. 
Please, specify.
>  it still doesn't say what tests and jobs need to be implemented
We have a huge feature coming. I am not going to predict, how many tests (and which exactly) will be enough to cover it.
> when we know the feature can be verified
Pass criteria is on line 27
> why do we need to execute this with test suite
validation on Openshift, see line 65

mjurc replied:
>  It seems a bit out of date.
Well, we're quite late into 3.27 cycle and none of this has been done while other checks not mentioned in this plan were implemented. So on one hand, what's written here doesn't exist and what's been done is not done according to the plan.

>  We have a huge feature coming. I am not going to predict, how many tests (and which exactly) will be enough to cover it.

Purpose of the test plan is decomposition and planning of work to make sure the feature is not verified by ad-hoc decisions. The feature was known to be huge for last two quarters, how comes no planning was really executed despite all the effort in reviews? This is the exact point of the test plan. Please update the test plan.
